### PR TITLE
WIP update to use InterTypes

### LINKED ISSUE #28: update SyntacticModels to use InterTypes

### PR DESCRIPTION
**Pull request by p-stokes**
This is to make the InterType form of the various structures currently in SyntacticModels.

Resolves #28

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -11,13 +11,15 @@ Decapodes = "679ab3ea-c928-4fe6-8d59-fd451142d391"
...
+OrderedCollections = "bac558e1-5e72-5ebc-8fee-abe8a469f55d"
 Reexport = "189a3867-3050-52da-a836-e630ba90ab69"
 StructTypes = "856f2bd8-1eba-4b0a-8007-ebc267875bd4"
+Unicode = "4ec0a83e-493e-50e2-b9ac-8f72acf5a8f5"
 
 [compat]
 ACSets = "0.2"
 Catlab = "0.15, 0.16"
-Decapodes = "0.5"
+Decapodes = "0.4, 0.5"
```
Comment (jpfairbanks) [inline Project.toml:22]: Since code is moved here, do we still need to depend on Decapodes?

File: src/amr_it.jl [hunk 2]
```diff
@@ -0,0 +1,391 @@
+module AMR
```
Comment (jpfairbanks) [inline src/amr_it.jl:1]: because of the rename of this file, the diff is useless. See above comment.

File: src/amr_it.jl [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+ Annotation, Note, Name, Description, Grounding, Units
+
+using Reexport
+@reexport using MLStyle
+@reexport using ACSets
+using ACSets.ADTs
+using ACSets.ACSetInterface
+using StructTypes
+
+# using ..SyntacticModelsBase
```
Comment (jpfairbanks) [inline src/amr_it.jl:17]: Delete commented out code

File: src/amr_it.jl [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+    amr.Normal(mu, var)   => "N($mu,$var)"
+    amr.PointMass(value)  => "δ($value)"
+  end
+end
+
+function distro_expr(d::amr.Distribution)
+  return Base.Meta.parse(distro_string(d))
+end
+
+#=
```
Comment (jpfairbanks) [inline src/amr_it.jl:39]: delete block comments.

File: src/amr_it.jl [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+      vs::Vector{amr.Pair}                 => map(vs) do v; "$(v[1]) => $(v[2])," end |> x-> join(x, "\n") 
+      vs::Vector{amr.Semantic}             => join(map(!, vs), "\n\n")
+      xs::Vector                       => map(!, xs)
+      # amr.Typing(system, map)              => "Typing = begin\n$(padlines(!system, 2))\nTypeMap = [\n$(padlines(!map, 2))]\nend"
+      # amr.ASKEModel(h, m, s)               => "$(!h)\n$(!m)\n\n$(!s)"
+      amr.Annotation(e,t,n)                => "Annotation = $(String(e)),$(String(t)): $(note_string(n))"
+    end
+  end
+end
+#=
```
Comment (jpfairbanks) [inline src/amr_it.jl:85]: why isn't this being used?

File: src/composite_models_it.jl [hunk 6]
```diff
@@ -0,0 +1,85 @@
+module Composites
```
Comment (jpfairbanks) [inline src/composite_models_it.jl:1]: again, I can't review this diff because of the file rename.

File: src/decapodes_it.jl [hunk 7]
```diff
@@ -0,0 +1,272 @@
+module ASKEMDecapodes
+
+export ASKEMDecaExpr, ASKEMDecapode, ASKEMDeca, SummationDecapode, parse_decapode
+
+# using ..SyntacticModelsBase
+using ..AMR
+
+using StructTypes
+# using Decapodes
```
Comment (jpfairbanks) [inline src/decapodes_it.jl:9]: is anything using Decapodes now that you copied code over here? Should we upgrade Decapodes to use InterTypes and then just use InterType inteheritance here?

File: src/decapodes_it.jl [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+      end
+      # TODO: delete unused vars. The only thing stopping me from doing 
+      # this is I don't know if CSet deletion preserves incident relations
+      #rem_parts!(d, :Var, sort(deletions))
+    end
+  end
+  return d
+end
+
+function recognize_types(d::decapodes.AbstractNamedDecapode)
```
Comment (jpfairbanks) [inline src/decapodes_it.jl:207]: Should SynacticModels be doing decapodes operations like `recognize_types`? What do you think the scope boundary is?

File: src/decapodes_it.jl [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+        Expr(:call, f, x, y) => decapodes.App2(f, term(x), term(y))
+
+        # TODO: Will later be converted to Op2's or schema has to be changed to include multiplication
+        Expr(:call, :*, xs...) => decapodes.Mult(term.(xs))
+
+        x => error("Cannot construct term from  $x")
+    end
+end
+
+function parse_decapode(expr::Expr)
```
Comment (jpfairbanks) [inline src/decapodes_it.jl:59]: Should SynacticModels be doing decapodes operations like `parse_decapode`? What do you think the scope boundary is?

File: src/uwd_it.jl [hunk 10]
```diff
@@ -0,0 +1,191 @@
+module ASKEMUWDs
```
Comment (jpfairbanks) [inline src/uwd_it.jl:1]: I can't review this diff because the file was renamed unnecessarily.

**Review by jpfairbanks — CHANGES_REQUESTED** [review]
The unnecessary rename of these file makes the diffs unreviewable. Please put them back and confirm that the diffs are useful. I can review again at that time.

**Comment by p-stokes** [comment]
> The unnecessary rename of these file makes the diffs unreviewable. Please put them back and confirm that the diffs are useful. I can review again at that time.

I've updated the file names for the PR, and the diffs look ok now. Regarding the included decapodes functions, I think the longer-term plan is for Decapodes and other structures to be defined directly upstream with InterTypes. But for the short-term, it seemed easier to develop locally within SyntacticModels.

I do think it would be helpful to have a discussion about the design/scope plan for SyntacticModels because it relates to the current points of development for the AMR and Composites functionality as well.

File: src/SyntacticModels.jl [hunk 11]
```diff
@@ -1,9 +1,9 @@
 module SyntacticModels
 
 include("SyntacticModelsBase.jl")
-include("amr.jl")
-include("decapodes.jl")
-include("uwd.jl")
-include("composite_models.jl")
+include("amr_it.jl")
```
[the remark this answers is not in the document]
Comment (p-stokes) [reply src/SyntacticModels.jl]: It was just for development purposed. I've renamed them.

File: test/composite_models_examples.jl [hunk 12]
```diff
@@ -5,33 +5,39 @@ using ..SyntacticModels.Composites
...
-  Statement(:heating, [v,Q])]
-u = ASKEMUWDs.UWDExpr(c, s)
+s = [ASKEMUWDs.uwd.Statement(:oscillator, [x,v]),
+  ASKEMUWDs.uwd.Statement(:heating, [v,Q])]
+u = UWDExpr(c, s)
 
 
 
-h = AMR.Header("harmonic_oscillator",
+h = Header("","harmonic_oscillator",
```
[the remark this answers is not in the document]
Comment (p-stokes) [reply test/composite_models_examples.jl:25]: In the Model-Representations repo, I noticed Five had added an extra `id` field to the `Header` there, so I incorporated it here. I can go through and take it out if you want?

File: test/composite_models_examples.jl [hunk 13]
```diff
@@ -5,33 +5,39 @@ using ..SyntacticModels.Composites
...
 using Catlab.RelationalPrograms
 using Catlab.WiringDiagrams
 using Test
 
+using ACSets
+using ACSets.InterTypes
+using Test
+using OrderedCollections
+import JSON
+import JSON3
```
[the remark this answers is not in the document]
Comment (p-stokes) [reply test/composite_models_examples.jl:11]: Similar to the name qualifications, I was a bit lax with the imports during development. 
I have already begun going through and cleaning up some of the unnecessary imports.

**Comment by p-stokes** [comment]
I cleaned up and standardized the imports and name qualifications. Still working on the remaining composite `oapply` example and amr `acsetspec` examples

File: src/composite_models.jl [hunk 14]
```diff
@@ -1,52 +1,66 @@
 module Composites
 
-export CompositeModelExpr, OpenModel, OpenDecapode, CompositeModel, interface, open_decapode
+export CompositeModelExpr, OpenModel, OpenDecapode, CompositeModel, interface, open_decapode, oapply, Open
 
 using MLStyle
 using Catlab
-using Decapodes
+using Decapodes # : SummationDecapode
```
[the remark this answers is not in the document]
Comment (p-stokes) [reply src/composite_models.jl]: It was deleted in the imports cleanup.

File: src/amr.it [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+  Value(_1::Symbol)
+  Kwarg(_1::Symbol, _2::Args)
+end
+
+struct Statement
+  table::Symbol
+  element::Vector{Args}
+end
+
+#=
```
Comment (jpfairbanks) [inline src/amr.it:12]: do we need this?

File: test/composite_models_examples.jl (hunk 12 above)
[the remark this answers is not in the document]
Comment (p-stokes) [reply test/composite_models_examples.jl:25]: Ok. I'll undo that and make a separate issue and PR after.

File: src/amr.it [hunk 16]
```diff
@@ -0,0 +1,102 @@
+
+@sum Args begin
```
Comment (jpfairbanks) [inline src/amr.it:2]: Is this being copied from https://github.com/AlgebraicJulia/ACSets.jl/blob/main/src/ADTs.jl#L14?

We should upgrade that code to use InterTypes and then import that usage here.

**Comment by p-stokes** [comment]
> > I do think it would be helpful to have a discussion about the design/scope plan for SyntacticModels because it relates to the current points of development for the AMR and Composites functionality as well.
> 
> I think the first part of that discussion is understanding your expectations for the scope. What are your thoughts?

Summarizing our in-person discussion about this: 
The short-term scope is to convert the functionality and test examples to use `InterTypes` locally within SyntacticModels.jl (as best makes sense and is possible). Longer-term, the .it files and functions will be upstreamed to their respective repositories.
There are several blocking issues with `InterTypes`' capabilities and integration with some other upstream functionality (e.g., `ADTs` and `Open`). I will make issues for those and tag them below.

**Comment by olynch** [comment]
Some of the work here ends up duplicating work merged into DiagrammaticEquations, so it makes sense to make a new PR

**Pull request closed without merging**